### PR TITLE
[Core] Support `min_tokens` with speculative decoding

### PR DESCRIPTION
**Pull request by qianlihuang**
## Purpose

Allow `min_tokens` in speculative decoding by masking stop tokens during rejection sampling.

## Test Plan

Run the server with speculative decoding enabled (example):

1. 
```sh
vllm serve deepseek-ai/DeepSeek-V3.2 \
		--tensor-parallel-size 8 \
		--tokenizer-mode deepseek_v32 \
		--tool-call-parser deepseek_v32 \
		--enable-auto-tool-choice \
		--reasoning-parser deepseek_v3 \
		--speculative_config '{"num_speculative_tokens":1, "method":"deepseek_mtp"}'
```

2.
```sh
CUDA_VISIBLE_DEVICES=0 vllm serve meta-llama/Llama-3.1-8B-Instruct     --tensor-parallel-size 1 --speculative_config '{"num_speculative_tokens":1, "method":"eagle3","model":"lmsys/SGLang-EAGLE3-Llama-3.1-8B-Instruct-SpecForge"}' --max-model-len 16384
```

Then exercise `min_tokens` in an API call:

1.
```sh
curl -X POST http://localhost:8080/v1/chat/completions \
	-H "Content-Type: application/json" \
	-d '{
		"model": "deepseek-ai/DeepSeek-V3.2",
		"messages": [
			{"role": "user", "content": "Hello, how are you?"}
		],
		"min_tokens": 512,
		"max_tokens": 640,
		"seed": 42
	}'
```

2.
(1)
```sh
curl -X POST http://localhost:8080/v1/chat/completions   -H "Content-Type: application/json"   -d '{
    "model": "meta-llama/Llama-3.1-8B-Instruct",
    "messages": [
      {"role": "user", "content": "Hello, how are you?"}
    ],
    "min_tokens": 256,
    "max_tokens": 512,
    "seed": 42
 }'
```
(2)
```sh
curl -X POST http://localhost:8080/v1/chat/completions   -H "Content-Type: application/json"   -d '{
    "model": "meta-llama/Llama-3.1-8B-Instruct",
    "messages": [
      {"role": "user", "content": "Hello, how are you?"}
    ],
    "min_tokens": 512,
    "max_tokens": 1024,
    "seed": 42
 }'
```

## Test Result
1.
```
{"id":"chatcmpl-9864bb85cec23fd2","object":"chat.completion","created":1768890340,"model":"deepseek-ai/DeepSeek-V3.2","choices":[{"index":0,"message":{"role":"assistant","content":"Hello! I'm doing great, thank you for asking! 😊 I'm here and ready to help you with whatever you need. How are you doing today? Is there anything I can assist you with? Whether it's a question, a task, or just a friendly chat—I'm all ears! 💬✨\n\nLooking forward to helping you out! 😊🌟💬💭📚🔍💡🎉👋😄🚀🎯💯❤️✨🎨🤖📝🌈📊🔧🎓📌🔎🧠📈🎶📸🌍💻🛠️🎁🌱📖🔬💬🎊🤝🌟💡✨😊🎯💬🔍📚🚀❤️🎨🤖🌈📊🎓📌🧠📈🎶📸🌍💻🛠️🎁🌱📖🔬💬🎊🤝# /r/place 2023\n\nThis project is a collection of tools for working with the 2023 /r/place dataset.\n\n## Data\n\nThe dataset is available [here](https://placedata.reddit.com/data/canvas-history/index.html).\n\n### Downloading\n\n`download.py` can be used to download the dataset.\n\n```sh\n# Download to the current directory\npython download.py\n\n# Download to a specific directory\npython download.py -o /path/to/directory\n\n# Download specific files\npython download.py -f 2023_place_canvas_history-000000000000.csv.gz -f 2023_place_canvas_history-000000000001.csv.gz\n```\n\nThe script will automatically resume partial downloads.\n\n## Dataset Schema\n\nThe dataset is a collection of gzipped CSV files. Each row represents a single pixel placement.\n\n| Column | Type | Description |\n| ------ | ---- | ----------- |\n| `timestamp` | `timestamp` | The time the pixel was placed (UTC). |\n| `user_id` | `string` | A unique identifier for the user who placed the pixel. |\n| `pixel_color` | `string` | The color of the pixel placed. |\n| `coordinate` | `string` | The coordinate of the pixel placed, in the format `x,y`. |\n\n### Color Mapping\n\nThe `pixel_color` column is a hex string representing the color of the pixel placed. The mapping from hex string to color name is as follows:\n\n| Hex | Color |\n| --- | ----- |\n| `#6D001A` | `dark red` |\n| `#BE0039` | `red` |\n| `#FF4500` | `orange` |\n| `#FFA800` | `yellow` |\n| `#FFD635` | `light yellow` |\n| `#FFF8B8` | `dark green` |\n| `#00A368` | `green` |\n| `#00CC78` | `light green` |\n| `#7","refusal":null,"annotations":null,"audio":null,"function_call":null,"tool_calls":[],"reasoning":null,"reasoning_content":null},"logprobs":null,"finish_reason":"length","stop_reason":null,"token_ids":null}],"service_tier":null,"system_fingerprint":null,"usage":{"prompt_tokens":10,"total_tokens":650,"completion_tokens":640,"prompt_tokens_details":null},"prompt_logprobs":null,"prompt_token_ids":null,"kv_transfer_params":null}
```

2.
(1)
```
{"id":"chatcmpl-9d56664d8b826244","object":"chat.completion","created":1770093143,"model":"meta-llama/Llama-3.1-8B-Instruct","choices":[{"index":0,"message":{"role":"assistant","content":"I'm just a computer program, so I don't have feelings, but thank you for asking. I'm functioning properly and ready to assist you with any questions or topics you'd like to discuss. How can I help you today? Is there something specific on your mind, or would you like to chat about a particular subject? I'm here to help. \n\nAlso, I can provide information on a wide range of topics, from science and history to entertainment and culture. If you're looking for something specific, feel free to ask, and I'll do my best to provide a helpful response. \n\nIs there anything in particular that you'd like to know or discuss? I'm all ears... or rather, all text. Let me know how I can help. \n\nIf you're just looking for some conversation, I'm happy to chat with you about your day, your interests, or anything else that's on your mind. I'm a good listener, and I'm here to help you explore your thoughts and ideas. So, what's on your mind? I'm here to listen and help in any way I can. \n\nLet me know if there's anything specific you'd like to talk about, or if you'd just like to chat and see where the conversation takes us. I'm here to help and provide information, and I'm always happy to engage in a fun and interesting conversation.","refusal":null,"annotations":null,"audio":null,"function_call":null,"tool_calls":[],"reasoning":null,"reasoning_content":null},"logprobs":null,"finish_reason":"stop","stop_reason":null,"token_ids":null}],"service_tier":null,"system_fingerprint":null,"usage":{"prompt_tokens":41,"total_tokens":324,"completion_tokens":283,"prompt_tokens_details":null},"prompt_logprobs":null,"prompt_token_ids":null,"kv_transfer_params":null}
```
(2)
```
{"id":"chatcmpl-b7e559c7c34bfd3d","object":"chat.completion","created":1770093152,"model":"meta-llama/Llama-3.1-8B-Instruct","choices":[{"index":0,"message":{"role":"assistant","content":"I'm just a computer program, so I don't have feelings, but thank you for asking. I'm functioning properly and ready to assist you with any questions or topics you'd like to discuss. How can I help you today? Is there something specific on your mind, or would you like to chat about a particular subject? I'm here to help. \n\nAlso, I can provide information on a wide range of topics, from science and history to entertainment and culture. If you're looking for something specific, feel free to ask, and I'll do my best to provide a helpful response. \n\nIs there anything in particular that you'd like to know or discuss? I'm all ears... or rather, all text. Let me know how I can help. \n\nIf you're just looking for some conversation, I'm happy to chat with you about your day, your interests, or anything else that's on your mind. I'm a good listener, and I'm here to help you explore your thoughts and ideas. So, what's on your mind? I'm here to listen and help in any way I can. \n\nLet me know if there's anything specific you'd like to talk about, or if you'd just like to chat and see where the conversation takes us. I'm here to help and provide information, and I'm always happy to engage in a fun and interesting conversation. \n\nHow can I assist you today? Do you have a particular question or topic in mind, or would you like to explore something new? I'm here to help and provide information on a wide range of topics. Let me know how I can assist you. \n\nIf you're not sure where to start, we could play a game, have a fun conversation, or explore a new topic together. I'm here to help and provide information in a way that's engaging and easy to understand. So, what do you say? Is there anything in particular that you'd like to know or discuss? I'm here to help and provide information on a wide range of topics. Let me know how I can assist you. \n\nLet's get started, and see where the conversation takes us. I'm here to help and provide information in a way that's engaging and easy to understand. So, what do you say? Is there anything in particular that you'd like to know or discuss? I'm here to help and provide information on a wide range of topics. Let me know how I can assist you. \n\nI'm here to help and provide information on a wide range of topics. If you're not sure where to start, we could play a game, have a fun conversation, or explore a new topic together. I'm here to help and provide information in a way that's engaging and easy to understand. So, what do you say? Is there anything in particular that you'd like to know or discuss? I'm here to help and provide information on a wide range of topics. Let me know how I can assist you. \n\nLet's get started, and see where the conversation takes us. I'm here to help and provide information in a way that's engaging and easy to understand. So, what do you say? Is there anything in particular that you'd like to know or discuss? I'm here to help and provide information on a wide range of topics. Let me know how I can assist you.","refusal":null,"annotations":null,"audio":null,"function_call":null,"tool_calls":[],"reasoning":null,"reasoning_content":null},"logprobs":null,"finish_reason":"stop","stop_reason":null,"token_ids":null}],"service_tier":null,"system_fingerprint":null,"usage":{"prompt_tokens":41,"total_tokens":731,"completion_tokens":690,"prompt_tokens_details":null},"prompt_logprobs":null,"prompt_token_ids":null,"kv_transfer_params":null}
```

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

File: vllm/v1/sample/metadata.py [hunk 1]
```diff
@@ -42,3 +42,7 @@ class SamplingMetadata:
 
     # Speculative token ids
     spec_token_ids: list[list[int]] | None = None
+
+    min_tokens: list[int] | None = None
+    all_stop_token_ids: list[set[int]] | None = None
```
Comment (benchislett) [inline vllm/v1/sample/metadata.py]: I don't like that we have to add all_stop_token_ids to the sampling metadata here. How is min_tokens handled for non-specdec cases, and can we reuse some logic/structure from that flow?

Comment (qianlihuang) [reply vllm/v1/sample/metadata.py]: @benchislett  Thanks for the review! I’ve updated it to reuse the existing MinTokensLogitsProcessor flow

File: vllm/v1/sample/rejection_sampler.py [hunk 2]
```diff
@@ -292,6 +292,15 @@ def apply_logits_processors(
                 logits, bad_words_token_ids, output_token_ids, metadata.num_draft_tokens
             )
 
+        # Apply min_tokens constraint via MinTokensLogitsProcessor.
+        from vllm.v1.sample.logits_processor.builtin import MinTokensLogitsProcessor
```
Comment (benchislett) [inline vllm/v1/sample/rejection_sampler.py]: Not sure why this import needs to be inline?

File: vllm/v1/sample/logits_processor/__init__.py [hunk 3]
```diff
@@ -202,10 +202,11 @@ def build_logitsprocs(
         if custom_logitsprocs:
             raise ValueError(STR_SPEC_DEC_REJECTS_LOGITSPROCS)
         logger.warning(
-            "min_p, logit_bias, and min_tokens parameters won't currently work "
-            "with speculative decoding enabled."
+            "min_p and logit_bias parameters won't work with speculative decoding."
+        )
+        return LogitsProcessors(
+            iter([MinTokensLogitsProcessor(vllm_config, device, is_pin_memory)])
```
Comment (benchislett) [inline vllm/v1/sample/logits_processor/__init__.py]: What's the point of having an `iter` here?

File: vllm/v1/sample/rejection_sampler.py [hunk 4]
```diff
@@ -292,6 +293,12 @@ def apply_logits_processors(
                 logits, bad_words_token_ids, output_token_ids, metadata.num_draft_tokens
             )
 
+        for processor in sampling_metadata.logitsprocs.non_argmax_invariant:
```
Comment (benchislett) [inline vllm/v1/sample/rejection_sampler.py:296]: Maybe naive question, but why do we apply non_argmax_invariant only here? AFAIK rejection sampling doesn't necessarily do greedy sampling for verification. Maybe the argmax_invariant ones are applied elsewhere?

Comment (qianlihuang) [reply vllm/v1/sample/rejection_sampler.py:296]: @benchislett 

The split here is by processor semantics (is_argmax_invariant()), not by whether verification is greedy.

Current built-in processors are MinTokensLogitsProcessor, LogitBiasLogitsProcessor, and MinPLogitsProcessor (vllm/v1/sample/logits_processor/init.py).

Their is_argmax_invariant() are: MinP=True, LogitBias=False, MinTokens=False (vllm/v1/sample/logits_processor/builtin.py).

Under speculative decoding, build_logitsprocs() takes an early-return path that constructs only MinTokensLogitsProcessor and rejects user custom processors, so argmax_invariant is empty by construction in this path.(vllm/v1/sample/logits_processor/init.py)

argmax_invariant processors are still applied in the regular sampler flow (vllm/v1/sample/sampler.py).

If/when we support MinP/LogitBias in spec-decode, we should add spec-decode-aware handling for the expanded draft-token logits layout.

File: vllm/v1/sample/logits_processor/__init__.py (hunk 3 above)
Comment (qianlihuang) [reply vllm/v1/sample/logits_processor/__init__.py]: Thanks! The iter() is unnecessary here.

File: vllm/v1/sample/logits_processor/builtin.py [hunk 5]
```diff
@@ -236,6 +237,51 @@ def apply(self, logits: torch.Tensor) -> torch.Tensor:
             logits.index_put_(self.logits_slice, self.neg_inf_tensor)
         return logits
 
+    def apply_with_spec_decode(
```
Comment (benchislett) [inline vllm/v1/sample/logits_processor/builtin.py:240]: It's a bit hard to follow how this code works, please add a few short comments, for instance:
- What is n_mask?
- What do all_toks and all_rows represent?
- What is logits_slice?

File: vllm/v1/sample/logits_processor/builtin.py [hunk 6]
```diff
@@ -236,6 +237,51 @@ def apply(self, logits: torch.Tensor) -> torch.Tensor:
...
+        if not self.min_toks:
+            return logits
+
+        num_draft_arr = np.array(num_draft_tokens, dtype=np.int64)
+        cumsum = np.concatenate([[0], np.cumsum(num_draft_arr)])
+
+        entries = [
+            (req_idx, min_tok, len(out_tok_ids), list(stop_tok_ids))
+            for req_idx, (min_tok, out_tok_ids, stop_tok_ids) in self.min_toks.items()
+            if stop_tok_ids and req_idx < len(num_draft_tokens)
```
Comment (benchislett) [inline vllm/v1/sample/logits_processor/builtin.py]: Is `req_idx < len(num_draft_tokens)` just a sanity check or are there cases where `req_idx` and be out-of-bounds here? If there are such cases, maybe it makes sense to assume no draft tokens but apply the min_toks anyways?

**Review by benchislett — APPROVED** [review]
It would probably be more efficient to do this processing in update_state, but that might mean we have to make assumptions about number of draft tokens for each request. This should do for now
